dotcom-rendering/src/web/components/DynamicFast.tsx

```diff
@@ -22,510 +30,731 @@ type Props = {
+		} else {
+			firstSliceLayout = 'twoVeryBigs';
+		}
+		firstSliceCards = groupedTrails.veryBig.slice(0, 2);
+		secondSliceGroupedTrails = {
+			...groupedTrails,
+			veryBig: groupedTrails.veryBig.slice(2),
+		};
 	}
 
-	/**
-	 * Bigs - Starting from the left of the second row
-	 *
-	 * Put together from (in order)
-	 *  - Any left over huges
-	 *  - Any left over very bigs
-	 *  - All bigs
-	 *
-	 * Limited to 4 as each big (unboosted) takes up 25% of the row
-	 */
-	const bigs = [
-		...groupedTrails.huge,
-		...groupedTrails.veryBig,
-		...groupedTrails.big,
-	].slice(0, 4);
+	let secondSliceLayout:
+		| 'oneBigBoosted'
+		| 'twoOrMoreBigsFirstBoosted'
+		| 'fourBigs'
+		| 'threeBigs'
+		| 'twoBigs'
+		| 'oneBig'
+		| 'noBigs';
 
-	const firstBigBoosted = !!bigs[0]?.isBoosted;
+	let secondSliceCards: TrailType[] = [];
+	const bigs = [
+		// Demote any left over 'huge' or 'veryBig' grouped cards
+		...secondSliceGroupedTrails.huge,
+		...secondSliceGroupedTrails.veryBig,
+		...secondSliceGroupedTrails.big,
+	];
 
-	/**
-	 * firstBigBoostedPlusBig is a layout used when the first big has been boosted,
-	 * and additional bigs have been added in fronts tool.
-	 *
-	 * When this happens, we switch to the FirstBigBoostedPlusBig component for rendering
-	 * the standards section of the second row.
-	 */
-	const firstBigBoostedPlusBig = firstBigBoosted && !!bigs[1];
+	switch (bigs.length) {
+		case 0: {
+			secondSliceLayout = 'noBigs';
+			secondSliceCards = [
+				...secondSliceGroupedTrails.standard.slice(0, 12),
+			];
+			break;
+		}
+		case 1: {
+			if (bigs[0].isBoosted) {
+				secondSliceLayout = 'oneBigBoosted';
+				secondSliceCards = [
+					...bigs.slice(0, 1),
+					...secondSliceGroupedTrails.standard.slice(0, 6),
+				];
+			} else {
+				secondSliceLayout = 'oneBig';
+				secondSliceCards = [
+					...bigs.slice(0, 1),
+					...secondSliceGroupedTrails.standard.slice(0, 9),
+				];
+			}
+			break;
+		}
+		case 2: {
+			if (bigs[0].isBoosted) {
+				secondSliceLayout = 'twoOrMoreBigsFirstBoosted';
+				secondSliceCards = [
+					...bigs,
+					...secondSliceGroupedTrails.standard.slice(0, 7),
+				];
+			} else {
+				secondSliceLayout = 'twoBigs';
+				secondSliceCards = [
+					...bigs.slice(0, 2),
+					...secondSliceGroupedTrails.standard.slice(0, 6),
+				];
+			}
+			break;
+		}
+		case 3: {
+			if (bigs[0].isBoosted) {
+				secondSliceLayout = 'twoOrMoreBigsFirstBoosted';
+				secondSliceCards = [
+					...bigs,
+					...secondSliceGroupedTrails.standard.slice(0, 6),
+				];
+			} else {
+				secondSliceLayout = 'threeBigs';
+				secondSliceCards = [
+					...bigs.slice(0, 3),
+					...secondSliceGroupedTrails.standard.slice(0, 3),
+				];
+			}
+			break;
+		}
+		default: {
+			if (bigs[0].isBoosted) {
+				secondSliceLayout = 'twoOrMoreBigsFirstBoosted';
+				secondSliceCards = [
+					...bigs,
+					...secondSliceGroupedTrails.standard.slice(0, 5),
+				];
+			} else {
+				secondSliceLayout = 'fourBigs';
```

Comment:
if `secondSliceLayout` doesn't get set in this switch block then it can create a runtime error when `SecondSlice` switches on `secondSliceLayout` below, because the latter switch doesn't have a default case.

`tsc` isn't helping us here (not sure if this is because the assignment is complex, or because `secondSliceLayout` is initialised as `undefined`?

I think the code as written works, but with such a complex set of conditions I feel a bit uneasy about not being able to rely on the compiler/tooling to tell us whether we're actually catching all of the different cases. To prevent the bug, we could just add a default case on line 752 in this file, but it would be nice if  we could restructure to allow the compiler to tell us when we risk not setting a value for `secondSliceLayout`. I realise that mutation is well-motivated here, but wondering if there's a way around it? (Or a different way of narrowing the type before it gets used?)

Reply:
+1 I've added a default case to `SecondSlice` - it's frustrating tsc doesn't pick up on this :(